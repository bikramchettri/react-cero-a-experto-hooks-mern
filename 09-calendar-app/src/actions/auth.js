import { types } from '../types/types';
import { fetchSinToken } from '../helpers/fetch';
import Swal from 'sweetalert2';

export const startLogin = ( email, password ) => { 

    return async ( dispatch ) => {
        const respuesta = await fetchSinToken('auth', { email, password }, 'POST');
        const body = await respuesta.json();

        if( body.ok ){
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name 
            }))
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    } 
}

export const startRegister = ( name, email, password ) => { 
    return async ( dispatch ) => {
        const respuesta = await fetchSinToken('auth/new', { name, email, password}, 'POST');
        const body = await respuesta.json();

        if( body.ok ){
            localStorage.setItem('token', body.token );
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name 
            }))
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

export const login = ( user ) => ({ 
    type: types.authLogin,
    payload: user
})



export const authChecking = ( event ) => ({
    type: types.authChecking,
    payload: event
})

export const authCheckingFinish = ( event ) => ({ 
    type: types.authCheckingFinish,
    payload: event
})




export const authStartTokenRenew = ( event ) => ({ 
    type: types.authStartTokenRenew,
    payload: event
})

export const authLogout = ( event ) => ({ 
    type: types.authLogout
})