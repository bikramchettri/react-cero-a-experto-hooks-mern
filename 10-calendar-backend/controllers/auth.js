
const { response } = require('express'); //requiero express para no perder el intellisense
const Usuario = require('../models/Usuario'); //requiero al esquema del modelo Usuario
const bcrytp = require('bcryptjs'); //para la encriptacion de contraseñas
const { generarJWT } = require('../helpers/jwt') 

const loginUsuario = async (req, res = response ) => {

    const { email, password } = req.body;

    try {
        const usuario = await Usuario.findOne({email}); //le objeto se podria igial dejarlo así {email: email}

        if( !usuario ){
            return res.status(400).json({
                ok: false,
                msg: 'El usuario con ese correo no existe'
            });
        } 

        //confirmar los passwords, compara el password que se recibe en la peticion con el password en la DB
        const validPassword = bcrytp.compareSync(password, usuario.password );

        if( !validPassword ){
            return res.status(400).json({
                ok: false,
                msg: 'Password incorrecto'
            });
        }

        //Genera nuestro JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(200).json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        });

    } catch (error) {

        console.log(error)
        res.status(500).json({
            ok: false,
            msg: 'Por favor hable con el admin',
        });
        
    }
}


const crearUsuario = async (req, res = response) => {

    const { name, email, password } = req.body;
    
    try {
        //findOne es una promesa, retorna un objeto que cumpla esa condicion, sino regresar un null
        let usuario = await Usuario.findOne({email}); //le objeto se podria igial dejarlo así {email: email}

        if(usuario){
            return res.status(400).json({
                ok: false,
                msg: 'Un usuario ya existe con ese correo'
            });
        } 

        usuario = new Usuario( req.body ); //se le pasa todo lo que se obtiene desde el body del request (name, email, password)
        
        //Encriptar contraseña
        const salt = bcrytp.genSaltSync(); //genero la sal
        usuario.password = bcrytp.hashSync( password, salt); //encripto la contraseña usando la sal

        await usuario.save(); //guardo en la DB
        //Genera el JWT 
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(201).json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        });
                
    } catch (error) {
        console.log(error)
        res.status(500).json({
            ok: false,
            msg: 'Por favor hable con el admin',
        }); 
    } 


    
}

const revalidarToken = (req, res = response) => {
    res.json({
        ok: true,
        msg: 'renew token'
    });
}

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}