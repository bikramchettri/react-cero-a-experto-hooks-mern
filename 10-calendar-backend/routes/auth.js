/** 
 Rutas de usuarios o Auth 
 host + /api/auth
*/
const { Router} = require('express'); //desestructuro de express y obtengo a Router
//const router = express.Router; 
const router = Router();
const { loginUsuario, crearUsuario, revalidarToken } = require('../controllers/auth');

router.post('/new', crearUsuario);
router.post('/', loginUsuario );
router.get('/renew', revalidarToken );


module.exports = router; //exporto, algo diferente a Ecma