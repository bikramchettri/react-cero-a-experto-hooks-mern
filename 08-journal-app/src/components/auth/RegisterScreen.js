import React from 'react';
import {useDispatch} from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';

import { useForm } from '../../hooks/useForms';
import { startLoginEmailPassword, startGoogleLogin } from '../../actions/auth';


export const RegisterScreen = () => {

    const dispatch = useDispatch();

    const [formValues, handleInputChange ] = useForm({
        name:'Luci',
        email: 'heymundo@gmail.com',
        password:'123456',
        password2:'123456'
    });

    const {name, email, password, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if( isFormValid()){
            console.log('formulario valido');  
        }
    }

    const isFormValid = () => {
        if(name.trim().length === 0){
            console.log('nombre es requerido');
            return false;
        } else if(!validator.isEmail(email)){
            console.log('email no es valido');
            return false;
        }else if(password !== password2 || password.length < 5 ){
            console.log('las contraseÃ±as deben de coincidir y que sea de al menos 5 caracteres');
            return false;
        }
        return true;
    }

    return (
        <>
            <h3 className="auth__title"> Register</h3>

            <form onSubmit={handleRegister}>

                <div className="auth___alert-error">
                    Hola mundo
                </div>

                <input
                    type="text"
                    placeholder="Nombre"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={name}
                    onChange={handleInputChange}
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={handleInputChange}
                />
                
                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                    value={password2}
                    onChange={handleInputChange}
                />



                <button
                    className="btn btn-primary btn-block mb-5"
                    type="submit"
                    

                >
                    Registrar
                </button>


                
                <Link
                    className="link"
                    to="/auth/login"
                >
                    Already registered?
                </Link>
            </form>
        </>
    )
}
