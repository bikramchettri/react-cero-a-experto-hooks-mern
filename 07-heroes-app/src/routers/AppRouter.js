import React, { useContext } from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route
} from 'react-router-dom';

import { PrivateRoute } from './PrivateRoute';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { AuthContext } from '../auth/AuthContext';

export const AppRouter = () => {
    
    const {user} = useContext(AuthContext); //obtengo los datos del usuario si est√° logueado
    
    return (
        <Router>
            <div>
            
                <Switch>
                    <Route exact path="/login" component={ LoginScreen }/>
                    <PrivateRoute
                        path="/"
                        component={ DashboardRoutes }
                        isAuthenticated={user.logged}
                        />
                </Switch>
            </div>
        </Router>
    )
}
