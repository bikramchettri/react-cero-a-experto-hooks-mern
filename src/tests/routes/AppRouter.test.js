import React  from 'react';
import {mount}  from 'enzyme';
import { AppRouter } from '../../routers/AppRouter';
import { AuthContext } from '../../auth/AuthContext';


describe('Pruebas en AppRouter', () => {

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged:false
        }
    }
   
    test('debe de mostrar el login sino está autenticado', () => {
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue  }>
                <AppRouter/>
            </AuthContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
   });

   test('Debe de mostrar el component Marvel si está autenticado', () => {
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged:true,
                name:'Pedro'
            }
        };

       const wrapper = mount(
            <AuthContext.Provider value={ contextValue  }>
                <AppRouter/>
            </AuthContext.Provider>
        );
        
        expect(wrapper.find('.navbar').exists()).toBe(true);
   })
   
    
});