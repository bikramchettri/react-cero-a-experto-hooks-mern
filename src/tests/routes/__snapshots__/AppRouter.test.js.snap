// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AppRouter debe de mostrar el login sino est√° autenticado 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      LoginScreen
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Ingresar
    </button>
  </div>
</div>
`;
